docs/tutorial/overview.md

```diff
@@ -28,33 +27,112 @@ component main() -> {
 }
 ```
 
-Once we parse our MrXL program into an [AST][astcode], we can decide on rules for generating code depending on which AST node we're working on. Depending on the AST node, we might need to add code to `cells`, `wires` or `control`.
+`cells` contains declarations for logical hardware units like adders, memories and registers. `wires` contains `group`s that connect together the units declared in `cell`s and form the structure of the hardware. `control` contains the logic specifying when the `group`s will perform their computation.
+
+First, we'll parse our program into an [AST][astcode]. This program:
+
+```
+input foo: int[4]
+output baz: int[4]
+baz := map 4 (a <- foo) { a + 5 }
+```
+
+has an AST that we could represent in Python like this:
+
+```
+Prog(
+  decls=[
+    Decl(input=True, name='foo', type=Type(base='Tree(int, [])', size=4)),
+    Decl(input=False, name='baz', type=Type(base='Tree(int, [])', size=4))
+  ],
+  stmts=[
+    Stmt(dest='baz',
+      op=Map(
+        par=4,
+        bind=[Bind(dest=['a'], src='foo')],
+        body=BinExpr(
+          op='add',
+          lhs=VarExpr(name='a'),
+          rhs=LitExpr(value=5)
+        )
+      )
+    )
+  ]
+)
+```
+
+Now we can decide on rules for generating code depending on which AST node we're working on. Depending on the AST node, we might need to add code to `cells`, `wires` or `control`.
 
 ### `Decl` nodes
 
-`Decl` nodes instantiate new memories and registers. Here's FuTIL code that creates a new memory `foo`:
+`Decl` nodes instantiate new memories and registers. Here's FuTIL code that creates a new memory `foo`, with 4 32-bit elements and a 32-bit indexor:
+
 ```
 foo = prim std_mem_d1(32, 4, 32);
 ```
 
-For each `Decl` node, we just determine if we're instantiating a memory or a register, and then translate that to a corresponding FuTIL declaration and place that inside the `cells` section of our generated program.
+For each `Decl` node, we need to determine if we're instantiating a memory or a register, and then translate that to a corresponding FuTIL declaration and place that inside the `cells` section of our generated program. To get you started, here's some code that walks through each register declaration and generates a FuTIL program with those registers:
+
```

Comment:
Again, using the mdbook reference feature to show some lines of code while hiding others would be useful here.